pkg/application/applyBlueprintSpecUseCase.go

```diff
@@ -0,0 +1,145 @@
+package application
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cloudogu/k8s-blueprint-operator/pkg/domain"
+	"github.com/cloudogu/k8s-blueprint-operator/pkg/domainservice"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type ApplyBlueprintSpecUseCase struct {
+	repo               domainservice.BlueprintSpecRepository
+	doguInstallUseCase doguInstallationUseCase
+	healthUseCase      ecosystemHealthUseCase
+}
+
+func NewApplyBlueprintSpecUseCase(
+	repo domainservice.BlueprintSpecRepository,
+	doguInstallUseCase doguInstallationUseCase,
+	healthUseCase ecosystemHealthUseCase,
+) *ApplyBlueprintSpecUseCase {
+	return &ApplyBlueprintSpecUseCase{
+		repo:               repo,
+		doguInstallUseCase: doguInstallUseCase,
+		healthUseCase:      healthUseCase,
+	}
+}
+
+func (useCase *ApplyBlueprintSpecUseCase) CheckEcosystemHealthUpfront(ctx context.Context, blueprintId string) error {
+	logger := log.FromContext(ctx).WithName("ApplyBlueprintSpecUseCase.CheckEcosystemHealthUpfront").
+		WithValues("blueprintId", blueprintId)
+
+	logger.Info("getting blueprint spec for checking ecosystem health")
+	blueprintSpec, err := useCase.repo.GetById(ctx, blueprintId)
+	if err != nil {
+		return fmt.Errorf("cannot load blueprint spec %q to check ecosystem health: %w", blueprintId, err)
+	}
+
+	healthResult, err := useCase.healthUseCase.CheckEcosystemHealth(ctx, blueprintSpec.Config.IgnoreDoguHealth)
+	if err != nil {
+		return fmt.Errorf("cannot check ecosystem health upfront of applying the blueprint %q: %w", blueprintId, err)
+	}
+	blueprintSpec.CheckEcosystemHealthUpfront(healthResult)
+
+	err = useCase.repo.Update(ctx, blueprintSpec)
+	if err != nil {
+		return fmt.Errorf("cannot save blueprint spec %q after checking the ecosystem health: %w", blueprintId, err)
+	}
+
+	return nil
+}
+
+func (useCase *ApplyBlueprintSpecUseCase) CheckEcosystemHealthAfterwards(ctx context.Context, blueprintId string) error {
+	logger := log.FromContext(ctx).WithName("ApplyBlueprintSpecUseCase.CheckEcosystemHealthAfterwards").
+		WithValues("blueprintId", blueprintId)
+
+	logger.Info("getting blueprint spec for checking ecosystem health afterwards")
+	blueprintSpec, err := useCase.repo.GetById(ctx, blueprintId)
+	if err != nil {
+		return fmt.Errorf("cannot load blueprint spec %q to check ecosystem health: %w", blueprintId, err)
+	}
+
+	healthResult, err := useCase.healthUseCase.WaitForHealthyEcosystem(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot check ecosystem health after applying the blueprint %q: %w", blueprintId, err)
+	}
+	blueprintSpec.CheckEcosystemHealthAfterwards(healthResult)
```

Comment:
Yov have to check ignoreDoguHealth here

Reply:
Afterwards with `ignoreDoguHealth=true` the dogu-operator cannot upgrade unavailable dogus. Is this intended?

Reply:
My interpretation of `ignoreDoguHealth=true` is that we apply the blueprint even if dogus are unhealthy. After applying the blueprint we do the health check again so that we can track when the blueprint upgrade is complete. If the ecosystem is not healthy before the timeout, we will deactivate the maintenance mode anyways and mark the blueprint as failed.
I thought that the blueprint could be used to fix the unhealthy dogus. Besides maybe marking the blueprint as failed instead of completed, ignoring this flag here makes no real difference.

Reply:
> Afterwards with `ignoreDoguHealth=true` the dogu-operator cannot upgrade unavailable dogus. Is this intended?

You mean because the Dogu Operator cannot call the pre-upgrade script in the unhealthy dogu?

Reply:
We decided to have a look at this in the future.